sdk/storage/azure-storage-files-shares/src/share_client.cpp

```diff
@@ -41,31 +41,44 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
       const ShareClientOptions& options)
       : m_shareUrl(shareUrl)
   {
-    Azure::Core::Http::TelemetryOptions telemetryPolicyOptions;
-    telemetryPolicyOptions.ApplicationId = options.ApplicationId;
+    ShareClientOptions newOptions = options;
+    newOptions.PerRetryPolicies.insert(
+        newOptions.PerRetryPolicies.begin(),
+        std::make_unique<Storage::Details::StoragePerRetryPolicy>());
+    newOptions.PerRetryPolicies.emplace_back(
+        std::make_unique<Storage::Details::SharedKeyPolicy>(credential));
+    {
+      Azure::Core::Http::Internal::ValueOptions valueOptions;
+      valueOptions.HeaderValues[Storage::Details::HttpHeaderXMsVersion] = newOptions.ApiVersion;
+      newOptions.PerOperationPolicies.insert(
+          newOptions.PerOperationPolicies.begin(),
+          std::make_unique<Azure::Core::Http::Internal::ValuePolicy>(valueOptions));
+    }
     m_pipeline = std::make_shared<Azure::Core::Internal::Http::HttpPipeline>(
-        Storage::Details::ConstructPolicies(
-            std::make_unique<Azure::Core::Http::TelemetryPolicy>(
-                Storage::Details::BlobServicePackageName,
-                Details::Version::VersionString(),
-                telemetryPolicyOptions),
-            std::make_unique<Storage::Details::SharedKeyPolicy>(credential),
-            options));
+        newOptions,
+        Storage::Details::FileServicePackageName,
+        Details::Version::VersionString(),
+        std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>>(),
+        std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>>());
```

Comment:
Here is where you would pass all the policies that you are inserting to the end-user policies. 
You might add the sharedKeyAuthPolicy to the end-user policies just to work-around the need of letting it be the second last, but any other policy should be here.

Reply:
> You might add the sharedKeyAuthPolicy to the end-user policies just to work-around

Yeah, and I have to make a copy of the client options. Then I hit the bug.

Reply:
Got it, You are creating the copy to avoid mutating the end-user's policy :) 
This copy will be expensive for you. I would first check if the end-user-per-retry policies array contains at least one policy, otherwise, you can just add the auth policy in service's per-retry array.

If there is an end-user policy, then make the copy.  I would expect that if you define a default copy and move constructor would be enough as it would call the base class constructor as default (I will try it).

Reply:
> This copy will be expensive for you. I would first check if the end-user-per-retry policies array contains at least one policy, otherwise, you can just add the auth policy in service's per-retry array.

This sounds like a premature optimization. Anyway we cannot avoid copying the whole client options if end-user per-retry policies are not empty.

Reply:
> I would expect that if you define a default copy and move constructor would be enough as it would call the base class constructor as default

I don't need to define a default copy/move constructor. The compiler helps me generate one, one that copies every fields in derived-class and also calls copy/move constructor of base class. But some fields aren't copied in the copy constructor of base class. This is the bug.